// @stryker-mutator/core/dist/src/initializer/presets/angular-preset.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AngularPreset = void 0;
const tslib_1 = require("tslib");
const os_1 = (0, tslib_1.__importDefault)(require("os"));
const guideUrl = 'https://stryker-mutator.io/docs/stryker-js/guides/angular';
class AngularPreset {
    constructor() {
        this.name = 'angular-cli';
        // Please keep config in sync with handbook
        this.dependencies = ['@stryker-mutator/karma-runner'];
        this.config = {
            mutate: ['src/**/*.ts', '!src/**/*.spec.ts', '!src/test.ts', '!src/environments/*.ts'],
            testRunner: 'karma',
            karma: {
                configFile: 'karma.conf.js',
                projectType: 'angular-cli',
                config: {
                    browsers: ['ChromeHeadless'],
                },
            },
            reporters: ['progress', 'clear-text', 'html'],
            concurrency: Math.floor(os_1.default.cpus().length / 2),
            // eslint-disable-next-line camelcase
            concurrency_comment: 'Recommended to use about half of your available cores when running stryker with angular',
            coverageAnalysis: 'perTest',
        };
    }
    async createConfig() {
        return { config: this.config, guideUrl, dependencies: this.dependencies };
    }
}
exports.AngularPreset = AngularPreset;
