// @stryker-mutator/instrumenter/dist/src/mutators/block-statement-mutator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.blockStatementMutator = void 0;
const core_1 = require("@babel/core");
exports.blockStatementMutator = {
    name: 'BlockStatement',
    *mutate(path) {
        if (path.isBlockStatement() && isValid(path)) {
            yield core_1.types.blockStatement([]);
        }
    },
};
function isValid(path) {
    return !isEmpty(path) && !isInvalidConstructorBody(path);
}
function isEmpty(path) {
    return !path.node.body.length;
}
/**
 * Checks to see if a statement is an invalid constructor body
 * @example
 * // Invalid!
 * class Foo extends Bar {
 *   constructor(public baz: string) {
 *     super(42);
 *   }
 * }
 * @example
 * // Invalid!
 * class Foo extends Bar {
 *   public baz = 'string';
 *   constructor() {
 *     super(42);
 *   }
 * }
 * @see https://github.com/stryker-mutator/stryker-js/issues/2314
 * @see https://github.com/stryker-mutator/stryker-js/issues/2474
 */
function isInvalidConstructorBody(blockStatement) {
    return !!(blockStatement.parentPath.isClassMethod() &&
        blockStatement.parentPath.node.kind === 'constructor' &&
        (containsTSParameterProperties(blockStatement.parentPath) || containsInitializedClassProperties(blockStatement.parentPath)) &&
        hasSuperExpressionOnFirstLine(blockStatement));
}
function containsTSParameterProperties(constructor) {
    return constructor.node.params.some((param) => core_1.types.isTSParameterProperty(param));
}
function containsInitializedClassProperties(constructor) {
    return (constructor.parentPath.isClassBody() &&
        constructor.parentPath.node.body.some((classMember) => core_1.types.isClassProperty(classMember) && classMember.value));
}
function hasSuperExpressionOnFirstLine(constructor) {
    return (core_1.types.isExpressionStatement(constructor.node.body[0]) &&
        core_1.types.isCallExpression(constructor.node.body[0].expression) &&
        core_1.types.isSuper(constructor.node.body[0].expression.callee));
}
