// @stryker-mutator/core/dist/src/unexpected-exit-handler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnexpectedExitHandler = void 0;
const di_1 = require("./di");
const signals = Object.freeze(['SIGABRT', 'SIGINT', 'SIGHUP', 'SIGTERM']);
class UnexpectedExitHandler {
    constructor(process) {
        this.process = process;
        this.unexpectedExitHandlers = [];
        this.processSignal = (_signal, signalNumber) => {
            // Just call 'exit' with correct exitCode.
            // See https://nodejs.org/api/process.html#process_signal_events, we should exit with 128 + signal number
            this.process.exit(128 + signalNumber);
        };
        this.handleExit = () => {
            this.unexpectedExitHandlers.forEach((handler) => handler());
        };
        process.on('exit', this.handleExit);
        signals.forEach((signal) => process.on(signal, this.processSignal));
    }
    registerHandler(handler) {
        this.unexpectedExitHandlers.push(handler);
    }
    dispose() {
        this.process.off('exit', this.handleExit);
        signals.forEach((signal) => this.process.off(signal, this.processSignal));
    }
}
exports.UnexpectedExitHandler = UnexpectedExitHandler;
UnexpectedExitHandler.inject = [di_1.coreTokens.process];
