// @stryker-mutator/jest-runner/dist/src/jest-plugins/mixin-jest-environment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mixinJestEnvironment = void 0;
const messaging_1 = require("../messaging");
function fullNameDescribeBlock(describe) {
    if (describe.parent) {
        const parentName = fullNameDescribeBlock(describe.parent);
        return `${parentName} ${describe.name}`.trimStart();
    }
    else {
        return ''; // describe.name === "ROOT_DESCRIBE_BLOCK"
    }
}
function fullName(test) {
    const suiteName = fullNameDescribeBlock(test.parent);
    return `${suiteName} ${test.name}`.trimStart();
}
const STRYKER_JEST_ENV = Symbol('StrykerJestEnvironment');
function mixinJestEnvironment(JestEnvironmentClass) {
    var _a;
    if (JestEnvironmentClass[STRYKER_JEST_ENV]) {
        return JestEnvironmentClass;
    }
    else {
        class StrykerJestEnvironment extends JestEnvironmentClass {
            constructor(config, context) {
                super(config, context);
                this.handleTestEvent = async (event, eventState) => {
                    var _b, _c;
                    // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
                    await ((_b = super.handleTestEvent) === null || _b === void 0 ? void 0 : _b.call(this, event, eventState));
                    if (messaging_1.state.coverageAnalysis === 'perTest' && event.name === 'test_start') {
                        const ns = (this.global[this.global.__strykerGlobalNamespace__] = (_c = this.global[this.global.__strykerGlobalNamespace__]) !== null && _c !== void 0 ? _c : {});
                        ns.currentTestId = fullName(event.test);
                    }
                };
                this.fileName = context.testPath;
            }
            async teardown() {
                var _b;
                const mutantCoverage = (_b = this.global[this.global.__strykerGlobalNamespace__]) === null || _b === void 0 ? void 0 : _b.mutantCoverage;
                messaging_1.state.handleMutantCoverage(this.fileName, mutantCoverage);
                await super.teardown();
            }
        }
        _a = STRYKER_JEST_ENV;
        StrykerJestEnvironment[_a] = true;
        return StrykerJestEnvironment;
    }
}
exports.mixinJestEnvironment = mixinJestEnvironment;
