// @stryker-mutator/core/dist/src/reporters/json-reporter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JsonReporter = exports.RESOURCES_DIR_NAME = void 0;
const tslib_1 = require("tslib");
const path_1 = (0, tslib_1.__importDefault)(require("path"));
const plugin_1 = require("@stryker-mutator/api/plugin");
const file_url_1 = (0, tslib_1.__importDefault)(require("file-url"));
const ReporterUtil = (0, tslib_1.__importStar)(require("./reporter-util"));
const INDENTION_LEVEL = 0;
exports.RESOURCES_DIR_NAME = 'strykerResources';
class JsonReporter {
    constructor(options, log) {
        this.options = options;
        this.log = log;
    }
    onMutationTestReportReady(report) {
        this.mainPromise = this.generateReport(report);
    }
    wrapUp() {
        return this.mainPromise;
    }
    async generateReport(report) {
        const filePath = path_1.default.normalize(this.options.jsonReporter.fileName);
        this.log.debug(`Using relative path ${filePath}`);
        await ReporterUtil.writeFile(path_1.default.resolve(filePath), JSON.stringify(report, null, INDENTION_LEVEL));
        this.log.info(`Your report can be found at: ${(0, file_url_1.default)(filePath)}`);
    }
}
exports.JsonReporter = JsonReporter;
JsonReporter.inject = (0, plugin_1.tokens)(plugin_1.commonTokens.options, plugin_1.commonTokens.logger);
