// @stryker-mutator/core/dist/src/reporters/progress-append-only-reporter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProgressAppendOnlyReporter = void 0;
const tslib_1 = require("tslib");
const os_1 = (0, tslib_1.__importDefault)(require("os"));
const progress_keeper_1 = require("./progress-keeper");
class ProgressAppendOnlyReporter extends progress_keeper_1.ProgressKeeper {
    onAllMutantsMatchedWithTests(matchedMutants) {
        super.onAllMutantsMatchedWithTests(matchedMutants);
        if (matchedMutants.length) {
            this.intervalReference = setInterval(() => this.render(), 10000);
        }
    }
    onAllMutantsTested() {
        if (this.intervalReference) {
            clearInterval(this.intervalReference);
        }
    }
    render() {
        process.stdout.write(`Mutation testing ${this.getPercentDone()} (elapsed: ${this.getElapsedTime()}, remaining: ${this.getEtc()}) ` +
            `${this.progress.tested}/${this.progress.total} tested (${this.progress.survived} survived, ${this.progress.timedOut} timed out)` +
            os_1.default.EOL);
    }
    getPercentDone() {
        return `${Math.floor((this.progress.tested / this.progress.total) * 100)}%`;
    }
}
exports.ProgressAppendOnlyReporter = ProgressAppendOnlyReporter;
