// @stryker-mutator/core/dist/src/checker/checker-child-process-proxy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CheckerChildProcessProxy = void 0;
const check_1 = require("@stryker-mutator/api/check");
const child_process_proxy_1 = require("../child-proxy/child-process-proxy");
const checker_worker_1 = require("./checker-worker");
class CheckerChildProcessProxy {
    constructor(options, loggingContext) {
        this.childProcess = child_process_proxy_1.ChildProcessProxy.create(require.resolve('./checker-worker'), loggingContext, options, {}, process.cwd(), checker_worker_1.CheckerWorker, options.checkerNodeArgs);
    }
    async dispose() {
        var _a;
        await ((_a = this.childProcess) === null || _a === void 0 ? void 0 : _a.dispose());
    }
    async init() {
        var _a;
        await ((_a = this.childProcess) === null || _a === void 0 ? void 0 : _a.proxy.init());
    }
    async check(mutant) {
        if (this.childProcess) {
            return this.childProcess.proxy.check(mutant);
        }
        return {
            status: check_1.CheckStatus.Passed,
        };
    }
}
exports.CheckerChildProcessProxy = CheckerChildProcessProxy;
